// custom-logging-and-alerts/index.js
// Will send a message to a sns topic that will send out a text, email, slack or whatever else we want to.
// We can bypass sns and just send to what we want but putting it into a sns topic allows us to quickly
// add or change who or what the alert message goes to 

const sendAlertMessage = async (message) => {
    const messageType = message.messageType
    console.log(`Sending alert message for logs to ${messageType} group`)
}

const formatLogMessage = (message) => {
    const err = new Error(message);
    const errorStack = err.stack.split("\n")[2];
    const functionName = errorStack.split(" ")[5];
    const regex = /\((.*):(\d+):(\d+)\)$/;
    const absolutePath = regex.exec(errorStack);
    const filename = absolutePath[1].split("/").pop();
    const logStackDetails = {
        fileName: filename,
        functionName: functionName,
        lineNumber: absolutePath[2],
        logMessage: err.message,
    }
    return logStackDetails
}

const error = async (message, statusCode) => {
    const logStackDetails = formatLogMessage(message)
    const {
        fileName,
        functionName,
        lineNumber,
        logMessage
    } = logStackDetails

    const formattedLogMessage = {
        messageType: "Error",
        fileName: fileName,
        functionName: functionName,
        lineNumber: lineNumber,
        message: logMessage,
        statusCode: statusCode || 400
    }

    sendAlertMessage(logMessage);
    return console.error(formattedLogMessage)
}

const warning = async (message, statusCode) => {
    const logStackDetails = formatLogMessage(message)
    const {
        fileName,
        functionName,
        lineNumber,
        logMessage
    } = logStackDetails

    const formattedLogMessage = {
        messageType: "Warning",
        fileName: fileName,
        functionName: functionName,
        lineNumber: lineNumber,
        message: logMessage,
        statusCode: statusCode || 200
    }

    sendAlertMessage(formattedLogMessage);
    return console.warn(formattedLogMessage)
}

const info = async (message, statusCode) => {
    const logStackDetails = formatLogMessage(message)
    const {
        fileName,
        functionName,
        lineNumber,
        logMessage
    } = logStackDetails

    const formattedLogMessage = {
        messageType: "Info",
        fileName: fileName,
        functionName: functionName,
        lineNumber: lineNumber,
        message: logMessage,
        statusCode: statusCode || 200
    }

    return console.info(formattedLogMessage)
}

module.exports.info = info;
module.exports.warning = warning;
module.exports.error = error;
